### PR TITLE
make be_exception_safe_for_storage_vectorized

### PR DESCRIPTION
**Pull request by laotan332**
make BE exception safe: for storage/vectorized

### CODE REVIEW

**Comment by starrocks-xueli** [comment]
please check all `obtain_` function.

**Comment by starrocks-xueli** [comment]
missed `tablet_rowset_reader` in `schema_change.cpp`.

**Review by starrocks-xupeng — DISMISSED** [review]

File: be/src/storage/vectorized/meta_reader.cpp [hunk 1]
```diff
@@ -125,9 +126,12 @@ Status MetaReader::_get_segments(const TabletSharedPtr& tablet, const Version& v
...
 
     std::vector<RowsetSharedPtr> rowsets;
-    tablet->obtain_header_rdlock();
-    Status acquire_rowset_st = tablet->capture_consistent_rowsets(_version, &rowsets);
-    tablet->release_header_lock();
+    Status acquire_rowset_st;
+    {
+        tablet->obtain_header_rdlock();
+        DeferOp deferop([&] { tablet->release_header_lock(); });
+        acquire_rowset_st = tablet->capture_consistent_rowsets(_version, &rowsets);
```
Comment (sduzh) [inline be/src/storage/vectorized/meta_reader.cpp:133]:
```
    std::shared_lock l(tablet->get_header_lock());
    acquire_rowset_st = tablet->capture_consistent_rowsets(_version, &rowsets);
```
I plan to remove the method `obtain_header_rdlock` in the future.

File: be/src/storage/vectorized/schema_change.cpp [hunk 2]
```diff
@@ -942,77 +944,84 @@ Status SchemaChangeHandler::_do_process_alter_tablet_v2_normal(const TAlterTable
...
-            RowsetSharedPtr rowset = new_tablet->get_rowset_by_version(version);
-            rowsets_to_delete.push_back(rowset);
+        new_tablet->obtain_push_lock();
+        DeferOp new_tablet_push_lock_release_guard([&]() { new_tablet->release_push_lock(); });
+
+        base_tablet->obtain_header_wrlock();
+        DeferOp base_tablet_header_wrlock_release_guard([&]() { base_tablet->release_header_lock(); });
+
+        new_tablet->obtain_header_wrlock();
+        DeferOp new_tablet_header_wrlock_release_guard([&]() { new_tablet->release_header_lock(); });
```
Comment (sduzh) [inline be/src/storage/vectorized/schema_change.cpp:964]: std::lock_guard l1(base_tablet->get_push_lock());
std::lock_guard l2(new_tablet->get_push_lock());
std::lock_guard l3(base_tablet->get_header_lock());
std::lock_guard l4(new_tablet->get_header_lock());

File: be/src/storage/vectorized/tablet_reader.cpp [hunk 3]
```diff
@@ -36,8 +37,8 @@ void TabletReader::close() {
 
 Status TabletReader::prepare() {
     _tablet->obtain_header_rdlock();
+    DeferOp release_lock([&] { _tablet->release_header_lock(); });
```
Comment (sduzh) [inline be/src/storage/vectorized/tablet_reader.cpp:40]: std::shared_lock l(_tablet->get_header_lock());

File: be/src/storage/vectorized/tablet_reader.cpp [hunk 4]
```diff
@@ -209,6 +210,7 @@ Status TabletReader::_init_delete_predicates(const TabletReaderParams& params, D
     PredicateParser pred_parser(_tablet->tablet_schema());
 
     _tablet->obtain_header_rdlock();
+    DeferOp release_lock([&] { _tablet->release_header_lock(); });
```
Comment (sduzh) [inline be/src/storage/vectorized/tablet_reader.cpp:213]: std::shared_lock l(_tablet->get_header_lock());

File: be/src/storage/vectorized/tablet_reader.cpp [hunk 5]
```diff
@@ -18,6 +18,7 @@
 #include "storage/vectorized/predicate_parser.h"
 #include "storage/vectorized/seek_range.h"
 #include "storage/vectorized/union_iterator.h"
+#include "util/defer_op.h"
```
Comment (sduzh) [inline be/src/storage/vectorized/tablet_reader.cpp:21]: This can be removed?

File: be/src/storage/vectorized/meta_reader.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #include "storage/rowset/segment_v2/column_reader.h"
 #include "storage/tablet.h"
 #include "storage/vectorized/chunk_helper.h"
+#include "util/defer_op.h"
```
Comment (sduzh) [inline be/src/storage/vectorized/meta_reader.cpp:13]: Remove this include.

**Review by sduzh — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**